File: runme/mcgen/local/datan/exasrc/e3lsq.f
      PROGRAM E3LSQ
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      PARAMETER (MAXN=20,MAXNR=3,N=20,NR=3)
      DIMENSION T(MAXN),Y(MAXN),DELTAY(MAXN),LIST(MAXNR),
     +X(MAXNR),CX(MAXNR,MAXNR),GX(MAXNR,MAXNR),A(MAXN,MAXNR)
      DOUBLE PRECISION LSQGSS
      EXTERNAL LSQGSS
C identify program to user
      WRITE(*,*)' Program E3LSQ demonstrates use of LSQNON'
      WRITE(*,*)' '
C create simulated data
      X(1)=1.0D0
      X(2)=1.2D0
      X(3)=0.4D0
      NSEED1=15
      NSEED2=215
      CALL RNE2IN(NSEED1,NSEED2)
      CALL RNSTNR(DELTAY,N)
      CALL RNECUY(Y,N)
      DELT=DBLE(2./21.)
      SIGMA=.1D0
      WRITE(*,*)'   T         Y       DELTAY'
      DO 10 I=1,N
        T(I)=DBLE(I)*DELT
        SIG=SIGMA*(DBLE(0.5)+Y(I))
        Y(I)=LSQGSS(X,NR,T(I))+DELTAY(I)*SIG
        DELTAY(I)=SIG
        WRITE(*,'(3F10.5)')T(I),Y(I),DELTAY(I)
10    CONTINUE
      WRITE(*,*)' '
C loop over different numbers of unfixed variables
      DO 50 NLOOP = 3,0,-1
        NRED=NLOOP
C set first approximation of unknowns and perform fit with LSQNON
        IF(NRED.EQ.3) THEN
          X(1)=.8D0
          X(2)=.8D0
          X(3)=.8D0
          LIST(1)=1
          LIST(2)=1
          LIST(3)=1
        ELSE IF(NRED.EQ.2) THEN
          X(1)=.8D0
          X(2)=1.2D0
          X(3)=.8D0
          LIST(1)=1
          LIST(2)=0
          LIST(3)=1
        ELSE IF(NRED.EQ.1) THEN
          X(1)=.8D0
          X(2)=1.2D0
          X(3)=.4D0
          LIST(1)=1
          LIST(2)=0
          LIST(3)=0
        ELSE IF(NRED.EQ.0) THEN
          X(1)=1.0D0
          X(2)=1.2D0
          X(3)=.4D0
          LIST(1)=0
          LIST(2)=0
          LIST(3)=0
        END IF
        WRITE(*,'(A,I2,A,I2,A,3I2,A,3F5.1)')' NR =',NR,
     +' , NRED = ',NRED,' , LIST = ',LIST,', first appr. X = ',X
        NSTEP=100
        CALL LSQNON(LSQGSS,T,Y,DELTAY,N,NR,NRED,LIST,X,CX,R,A,GX,
     +  NSTEP)
C output results
        WRITE(*,*)' Result of fit with LSQNON:'
        WRITE(*,'(A,F10.5,A,I5)')' Fit to Gaussian, M = ',R,
     +  ' , NSTEP = ',NSTEP
        WRITE(*,'(A,(3F10.5))')' X = ',(X(K),K=1,NR)
        IF(NRED.GT.0) THEN
          WRITE(*,*)' Covariance matrix CX ='
          CALL MTXWRT(CX,NRED,NRED)
          WRITE(*,*)' '
        END IF
50    CONTINUE
      END

File: runme/mcgen/local/datan/exasrc/e3lsq.c
#include <stdio.h>
#include "datsrc.h"

/* Table of constant values */

static integer c__3 = 3, c__20 = 20;

main(void)
{
    static integer i, k, nstep, nseed1, nseed2, nloop, nred, nr = 3, list[3];
    static double a[60] /* was [20][3] */, cx[9] /* was [3][3] */,
                  gx[9] /* was [3][3] */, r, t[20], x[3], y[20], sigma,
                  deltay[20], delt, sig;

/* identify program to user */
    printf("%s\n\n", "Program E3LSQ demonstrates use of LSQNON.");
/* create simulated data */
    x[0] = 1.;
    x[1] = 1.2;
    x[2] = .4;
    nseed1 = 15;
    nseed2 = 215;
    rne2in_(&nseed1, &nseed2);
    rnstnr_(deltay, &c__20);
    rnecuy_(y, &c__20);
    delt = .09523809523809523;
    sigma = .1;
    printf("%s\n", "   T         Y       DELTAY");
    for (i = 1; i <= 20; ++i) {
	t[i - 1] = (double) i * delt;
	sig = sigma * (.5 + y[i - 1]);
	y[i - 1] = lsqgss_(x, &c__3, &t[i - 1]) + deltay[i - 1] * sig;
	deltay[i - 1] = sig;
        printf("%10.5lf%10.5lf%10.5lf\n", t[i-1], y[i-1], deltay[i-1]);
    }
    printf("\n");
/* loop over different numbers of unfixed variables */
    for (nloop = 3; nloop >= 0; --nloop) {
        nred = nloop;
/* set first approximation of unknowns and perform fit with LSQNON */
	if (nred == 3) {
	    x[0] = .8;
	    x[1] = .8;
	    x[2] = .8;
	    list[0] = 1;
	    list[1] = 1;
	    list[2] = 1;
	} else if (nred == 2) {
	    x[0] = .8;
	    x[1] = 1.2;
	    x[2] = .8;
	    list[0] = 1;
	    list[1] = 0;
	    list[2] = 1;
	} else if (nred == 1) {
	    x[0] = .8;
	    x[1] = 1.2;
	    x[2] = .4;
	    list[0] = 1;
	    list[1] = 0;
	    list[2] = 0;
	} else if (nred == 0) {
	    x[0] = 1.;
	    x[1] = 1.2;
	    x[2] = .4;
	    list[0] = 0;
	    list[1] = 0;
	    list[2] = 0;
	}
        printf("%s%2li%s%2li%s%", "NR =", nr, " , NRED = ", nred,
               " , LIST = ");
        for (k = 1; k <= nr; ++k)
           printf("%2li", list[k-1]);
        printf("%s", ", first appr. X = ");
        for (k = 1; k <= nr; ++k)
           printf("%5.1lf", x[k-1]);
        printf("\n");
	nstep = 100;
	lsqnon_(lsqgss_, t, y, deltay, &c__20, &c__3, &nred, list, x, 
	       cx, &r, a, gx, &nstep);
/* output results */
        printf("%s\n", "Result of fit with LSQNON:");
        printf("%s%10.5lf%s%5li\n", "Fit to Gaussian, M = ", r,
               " , NSTEP = ", nstep);
        printf("%s", "X = ");
        for (k = 1; k <= nr; ++k)
            printf("%10.5lf", x[k-1]);
        printf("\n");
	if (nred > 0) {
            printf("%s\n", "Covariance matrix CX =");
	    mtxwrt_(cx, &nred, &nred);
            printf("\n");
	}
    }
    return 0;
} /* main */

